src/api/che-login-manager.ts

```diff
@@ -0,0 +1,529 @@
+
+export interface RefreshTokenLoginRecord {
+  refreshToken: string
+}
+
+export interface OcUserTokenLoginRecord {
+  subjectToken: string
+  subjectIssuer: string
+}
+
+export interface PasswordLoginRecord {
+  username: string
+  password: string
+}
+
+export function isRefreshTokenLoginData(loginData: LoginRecord): loginData is RefreshTokenLoginRecord {
+  return !!(loginData as RefreshTokenLoginRecord).refreshToken
+}
+
+export function isOcUserTokenLoginData(loginData: LoginRecord): loginData is OcUserTokenLoginRecord {
+  return !!(loginData as OcUserTokenLoginRecord).subjectToken
+}
+
+export function isPasswordLoginData(loginData: LoginRecord): loginData is PasswordLoginRecord {
+  return !!(loginData as PasswordLoginRecord).password
+}
+
+// Response structure from <che-host>/api/keycloak/settings
+interface CheKeycloakSettings {
+  'che.keycloak.logout.endpoint': string
+  'che.keycloak.jwks.endpoint': string
+  'che.keycloak.token.endpoint': string
+  'che.keycloak.userinfo.endpoint': string
+  'che.keycloak.client_id': string
+  'che.keycloak.username_claim': string
+  'che.keycloak.js_adapter_url': string
+  'che.keycloak.use_nonce': string
+
+  'che.keycloak.profile.endpoint'?: string
+  'che.keycloak.auth_server_url'?: string
+  'che.keycloak.password.endpoint'?: string
+  'che.keycloak.realm'?: string
+
+  'che.keycloak.oidc_provider'?: string
+  'che.keycloak.github.endpoint'?: string
+}
+
+// Response structure from Keycloak get access token endpoint
+interface KeycloakAuthTokenResponse {
+  access_token: string
+  expires_in: number
+  refresh_token: string
+  refresh_expires_in: number
+  token_type: string
+  scope: string
+}
+
+const REQUEST_TIMEOUT_MS = 10000
+const LOGIN_DATA_FILE_NAME = 'che-login-config.json'
+
+let loginContext: CheServerLoginManager | undefined
+/**
+ * Che server login sessions manager. Singleton.
+ * Uses refresh tokens for authentication.
+ * Usually, just using of getLoginData function is suitable.
+ */
+export class CheServerLoginManager {
+  private loginData: CheServerLoginConfig
+  private apiUrl: string
+  private username: string
+
+  private readonly dataFilePath: string
+  private readonly axios: AxiosInstance
+
+  private constructor(dataFilePath: string) {
+    this.dataFilePath = dataFilePath
+
+    this.loginData = {}
+    this.readLoginData()
+    this.apiUrl = this.loginData.lastLoginUrl || ''
+    this.username = this.loginData.lastUserName || ''
+
+    // Remove outdated login records
+    this.removeExpiredLogins()
+
+    // Make axios ignore untrusted certificate error for self-signed certificate case.
+    const httpsAgent = new https.Agent({ rejectUnauthorized: false })
+    this.axios = axios.create({
+      httpsAgent
+    })
+  }
+
+  /**
+   * Returns Che server login sessions manager.
+   * @param configDirPath path to chectl config folder
+   */
+  static async getInstance(configDirPath: string): Promise<CheServerLoginManager> {
```

Comment:
What is the benefits of using getInstance instead of constructor ?

Reply:
1. Reusing for the same cluster
2. Asynchronous operations allowed, which is required to finish object construction.